In [1]:
import pandas as pd
import numpy as np
import os

In [9]:
df = pd.read_csv('C:/Users/dell g3/Downloads/Fraud.csv')
df.head(10)

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815      170136.00       160296.36   
1     1   PAYMENT   1864.28  C1666544295       21249.00        19384.72   
2     1  TRANSFER    181.00  C1305486145         181.00            0.00   
3     1  CASH_OUT    181.00   C840083671         181.00            0.00   
4     1   PAYMENT  11668.14  C2048537720       41554.00        29885.86   
5     1   PAYMENT   7817.71    C90045638       53860.00        46042.29   
6     1   PAYMENT   7107.77   C154988899      183195.00       176087.23   
7     1   PAYMENT   7861.64  C1912850431      176087.23       168225.59   
8     1   PAYMENT   4024.36  C1265012928        2671.00            0.00   
9     1     DEBIT   5337.77   C712410124       41720.00        36382.23   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0            0.00        0               0  
1  M2044282225             0.0            0.00        0               0  
2   C553264065             0.0            0.00        1               0  
3    C38997010         21182.0            0.00        1               0  
4  M1230701703             0.0            0.00        0               0  
5   M573487274             0.0            0.00        0               0  
6   M408069119             0.0            0.00        0               0  
7   M633326333             0.0            0.00        0               0  
8  M1176932104             0.0            0.00        0               0  
9   C195600860         41898.0        40348.79        0               0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


### Checking for Null values

In [8]:
df.any().isnull()

step              False
type              False
amount            False
nameOrig          False
oldbalanceOrg     False
newbalanceOrig    False
nameDest          False
oldbalanceDest    False
newbalanceDest    False
isFraud           False
isFlaggedFraud    False
dtype: bool

### Checking for the distributions of fraud and non-fraud data

In [13]:
df['isFraud'].value_counts()

0    6354407
1       8213
Name: isFraud, dtype: int64

In [14]:
df['isFlaggedFraud'].value_counts()

0    6362604
1         16
Name: isFlaggedFraud, dtype: int64

#### The dataset is highley unbalanced so we should split it to Fraud and Legit

In [20]:
df[df['isFlaggedFraud'] == 1]

step      type       amount     nameOrig  oldbalanceOrg  \
2736446   212  TRANSFER   4953893.08   C728984460     4953893.08   
3247297   250  TRANSFER   1343002.08  C1100582606     1343002.08   
3760288   279  TRANSFER    536624.41  C1035541766      536624.41   
5563713   387  TRANSFER   4892193.09   C908544136     4892193.09   
5996407   425  TRANSFER  10000000.00   C689608084    19585040.37   
5996409   425  TRANSFER   9585040.37   C452586515    19585040.37   
6168499   554  TRANSFER   3576297.10   C193696150     3576297.10   
6205439   586  TRANSFER    353874.22  C1684585475      353874.22   
6266413   617  TRANSFER   2542664.27   C786455622     2542664.27   
6281482   646  TRANSFER  10000000.00    C19004745    10399045.08   
6281484   646  TRANSFER    399045.08   C724693370    10399045.08   
6296014   671  TRANSFER   3441041.46   C917414431     3441041.46   
6351225   702  TRANSFER   3171085.59  C1892216157     3171085.59   
6362460   730  TRANSFER  10000000.00  C2140038573    17316255.05   
6362462   730  TRANSFER   7316255.05  C1869569059    17316255.05   
6362584   741  TRANSFER   5674547.89   C992223106     5674547.89   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
2736446      4953893.08   C639921569             0.0             0.0        1   
3247297      1343002.08  C1147517658             0.0             0.0        1   
3760288       536624.41  C1100697970             0.0             0.0        1   
5563713      4892193.09   C891140444             0.0             0.0        1   
5996407     19585040.37  C1392803603             0.0             0.0        1   
5996409     19585040.37  C1109166882             0.0             0.0        1   
6168499      3576297.10   C484597480             0.0             0.0        1   
6205439       353874.22  C1770418982             0.0             0.0        1   
6266413      2542664.27   C661958277             0.0             0.0        1   
6281482     10399045.08  C1806199534             0.0             0.0        1   
6281484     10399045.08  C1909486199             0.0             0.0        1   
6296014      3441041.46  C1082139865             0.0             0.0        1   
6351225      3171085.59  C1308068787             0.0             0.0        1   
6362460     17316255.05  C1395467927             0.0             0.0        1   
6362462     17316255.05  C1861208726             0.0             0.0        1   
6362584      5674547.89  C1366804249             0.0             0.0        1   

         isFlaggedFraud  
2736446               1  
3247297               1  
3760288               1  
5563713               1  
5996407               1  
5996409               1  
6168499               1  
6205439               1  
6266413               1  
6281482               1  
6281484               1  
6296014               1  
6351225               1  
6362460               1  
6362462               1  
6362584               1

### Since isFlaggedFraud is equal to 1 in the cases of isFraud equals 1 so we may drop it

In [29]:
df = df.drop('isFlaggedFraud', axis=1)
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

   nameDest  oldbalanceDest  newbalanceDest  isFraud  
0  Merchant             0.0             0.0        0  
1  Merchant             0.0             0.0        0  
2  Customer             0.0             0.0        1  
3  Customer         21182.0             0.0        1  
4  Merchant             0.0             0.0        0

### Changing the nameDest column values to Merchant and Customer for further processing

In [26]:
mask = df['nameDest'].str.startswith('C')
df.loc[mask, 'nameDest'] = 'Customer'

In [27]:
mask = df['nameDest'].str.startswith('M')
df.loc[mask, 'nameDest'] = 'Merchant'

In [30]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

   nameDest  oldbalanceDest  newbalanceDest  isFraud  
0  Merchant             0.0             0.0        0  
1  Merchant             0.0             0.0        0  
2  Customer             0.0             0.0        1  
3  Customer         21182.0             0.0        1  
4  Merchant             0.0             0.0        0

### Spliting the dataset to Fraud and Legit

In [31]:
Legit = df[df.isFraud == 0]
Fraud = df[df.isFraud == 1]

### Exploring the Legit and Fraud sets

In [40]:
Fraud.shape, Legit.shape

((8213, 10), (6354407, 10))

In [37]:
Fraud['type'].value_counts()

CASH_OUT    4116
TRANSFER    4097
Name: type, dtype: int64

#### Fraud is occuring when the type of transaction is CASH_OUT or TRANSFER only

In [44]:
Fraud.amount.describe()

count    8.213000e+03
mean     1.467967e+06
std      2.404253e+06
min      0.000000e+00
25%      1.270913e+05
50%      4.414234e+05
75%      1.517771e+06
max      1.000000e+07
Name: amount, dtype: float64

In [45]:
Legit.amount.describe()

count    6.354407e+06
mean     1.781970e+05
std      5.962370e+05
min      1.000000e-02
25%      1.336840e+04
50%      7.468472e+04
75%      2.083648e+05
max      9.244552e+07
Name: amount, dtype: float64

In [48]:
df.groupby('isFraud').mean()

step        amount  oldbalanceOrg  newbalanceOrig  \
isFraud                                                            
0        243.235663  1.781970e+05   8.328287e+05   855970.228109   
1        368.413856  1.467967e+06   1.649668e+06   192392.631836   

         oldbalanceDest  newbalanceDest  
isFraud                                  
0          1.101421e+06    1.224926e+06  
1          5.442496e+05    1.279708e+06

### Using Under-Sampling technique on Legit data 

In [49]:
# Number of rows in Fraud data is = 8213
sample_legit = Legit.sample(n=8213)

### Adding the two datasets together

In [52]:
new_df = pd.concat([sample_legit, Fraud], axis=0)

In [53]:
new_df.shape

(16426, 10)

In [56]:
new_df.head()

step      type      amount     nameOrig  oldbalanceOrg  \
4932722   350  TRANSFER  1031585.12  C1927953648         5573.0   
4961630   351   PAYMENT    17710.68   C378251185            0.0   
1618179   157  TRANSFER   747630.71   C665548678        53328.0   
4937211   350   PAYMENT    34104.55  C1246007656        62454.0   
1653953   158  CASH_OUT    80966.91   C272676419            0.0   

         newbalanceOrig  nameDest  oldbalanceDest  newbalanceDest  isFraud  
4932722            0.00  Customer       481278.77      1512863.88        0  
4961630            0.00  Merchant            0.00            0.00        0  
1618179            0.00  Customer      1215482.95      1963113.66        0  
4937211        28349.45  Merchant            0.00            0.00        0  
1653953            0.00  Customer       483677.12       564644.03        0

In [57]:
new_df['isFraud'].value_counts()

0    8213
1    8213
Name: isFraud, dtype: int64

In [58]:
new_df.groupby('isFraud').mean()

step        amount  oldbalanceOrg  newbalanceOrig  \
isFraud                                                            
0        241.658225  1.840483e+05   8.014142e+05   822858.769670   
1        368.413856  1.467967e+06   1.649668e+06   192392.631836   

         oldbalanceDest  newbalanceDest  
isFraud                                  
0          1.127034e+06    1.259065e+06  
1          5.442496e+05    1.279708e+06

#### The mean of the new dataset is similar to the old one so the sampling split is good to continue

### Using One-hot-encodings technique on type and nameDest columns

In [59]:
one_hot = pd.get_dummies(new_df['type'], prefix='type')
new_df = pd.concat([new_df, one_hot], axis=1)
new_df = new_df.drop('type', axis=1)

In [60]:
one_hot = pd.get_dummies(new_df['nameDest'], prefix='nameDest')
new_df = pd.concat([new_df, one_hot], axis=1)
new_df = new_df.drop('nameDest', axis=1)

In [61]:
new_df.head()

step      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
4932722   350  1031585.12  C1927953648         5573.0            0.00   
4961630   351    17710.68   C378251185            0.0            0.00   
1618179   157   747630.71   C665548678        53328.0            0.00   
4937211   350    34104.55  C1246007656        62454.0        28349.45   
1653953   158    80966.91   C272676419            0.0            0.00   

         oldbalanceDest  newbalanceDest  isFraud  type_CASH_IN  type_CASH_OUT  \
4932722       481278.77      1512863.88        0             0              0   
4961630            0.00            0.00        0             0              0   
1618179      1215482.95      1963113.66        0             0              0   
4937211            0.00            0.00        0             0              0   
1653953       483677.12       564644.03        0             0              1   

         type_DEBIT  type_PAYMENT  type_TRANSFER  nameDest_Customer  \
4932722           0             0              1                  1   
4961630           0             1              0                  0   
1618179           0             0              1                  1   
4937211           0             1              0                  0   
1653953           0             0              0                  1   

         nameDest_Merchant  
4932722                  0  
4961630                  1  
1618179                  0  
4937211                  1  
1653953                  0

### Dropping unnecessary columns (step, nameOrig)

In [62]:
new_df = new_df.drop('step', axis=1)
new_df = new_df.drop('nameOrig', axis=1)
new_df.shape

(16426, 13)

### Using Standard Scaler on the highly distributed colum values

In [67]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
col_to_scale = ['amount', 'oldbalanceOrg', 'newbalanceOrig', 'oldbalanceDest', 'newbalanceDest']
new_df[col_to_scale] = scaler.fit_transform(new_df[col_to_scale])
new_df.head()

amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
4932722  0.109873      -0.378090       -0.205960       -0.101743   
4961630 -0.432003      -0.379817       -0.205960       -0.239925   
1618179 -0.041889      -0.363290       -0.205960        0.109058   
4937211 -0.423241      -0.360461       -0.194458       -0.239925   
1653953 -0.398195      -0.379817       -0.205960       -0.101054   

         newbalanceDest  isFraud  type_CASH_IN  type_CASH_OUT  type_DEBIT  \
4932722        0.062660        0             0              0           0   
4961630       -0.326681        0             0              0           0   
1618179        0.178533        0             0              0           0   
4937211       -0.326681        0             0              0           0   
1653953       -0.181368        0             0              1           0   

         type_PAYMENT  type_TRANSFER  nameDest_Customer  nameDest_Merchant  
4932722             0              1                  1                  0  
4961630             1              0                  0                  1  
1618179             0              1                  1                  0  
4937211             1              0                  0                  1  
1653953             0              0                  1                  0

### Splitting the data into features and label

In [69]:
X = new_df.drop('isFraud', axis=1)
y = new_df['isFraud']

In [73]:
X.shape, y.shape

((16426, 12), (16426,))

### Splitting the data into train, test splits with 80:20 ratio

In [75]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [76]:
X_train.shape, X_test.shape

((13140, 12), (3286, 12))

### Model Training 

In [85]:
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [80]:
# Majing a pipeline to train the data on different models
pipelines = {
    'lr':make_pipeline(LogisticRegression()),
    'rf':make_pipeline(RandomForestClassifier()),
    'svm':make_pipeline(SVC()),
}

In [81]:
fit_models = {}
for algo , pipeline in pipelines.items():
    model = pipeline.fit(X_train , y_train)
    fit_models[algo] = model

D:\downloads\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Model Evaluation

In [87]:
for algo , model in fit_models.items():
    pred = model.predict(X_test)
    print(algo , accuracy_score(y_test , pred) * 100)

lr 92.30066950699938
rf 99.39135727328059
svm 90.26171637248936


#### The Random Forest model has the highest accuracy of 99.39 %

In [90]:
# Evaluating the precision, recall, and F1-score for the best model (random forest)
for algo , model in fit_models.items():
    if algo == 'rf':
        
        pred = model.predict(X_test)
        print(classification_report(y_test , pred))

              precision    recall  f1-score   support

           0       1.00      0.99      0.99      1643
           1       0.99      1.00      0.99      1643

    accuracy                           0.99      3286
   macro avg       0.99      0.99      0.99      3286
weighted avg       0.99      0.99      0.99      3286



#### The precision, recall, and f1-score are 99% which indicates that this is a good classification model.

## Answering some questions 

1. Data cleaning including missing values, outliers and multi-collinearity. 
> Already done above.
2. Describe your fraud detection model in elaboration. 
> I implemented a model to calassify legit and fraud transactions based on a set of variables that I selected based on the information from the dataset.
3. How did you select variables to be included in the model?
> I selected the variables based on their correlation to the fraud and my intuition when I went through and read the data.
4. Demonstrate the performance of the model by using best set of tools. 
> I used accuracy, precision, recall, and F1-score as performance evaluators.
5. What are the key factors that predict fraudulent customer? 
> type of transaction, amount of money transfered, the destination account name (nameDest), the amount of money in the original account after the transaction (newbalanceOrig). 
6. Do these factors make sense? If yes, How? If not, How not? 
> Yes, because if the type is Transfer or Cash_out, and the amount of money is huge, and the destination account is not a merchant. and the amount of money in the original account is 0 after the transaction then it is more likely to be a fradulent transaction.
7. What kind of prevention should be adopted while company update its infrastructure?
> Implement a verification process that prevents any transfers of cash-outs before contacting the original customer and asking if he/she tried to make that transaction, or make a two-step-verification process before making such transaction with certain amount of money, or making the customer answer for two personal questions that were saved when making the account before making such transactions.
8. Assuming these actions have been implemented, how would you determine if they work?
> I would collect the data after some period of time (1 to 3 months) and check for the rate of fradulent transactions againts the old data before implementing these actions to see if the rate is lower or if it is still the same then I will try to take a different approach to prevent such fradulent transactions.